### PR TITLE
fix #19266; allow reassign discriminant field

### LINKED ISSUE #19266: Manual example: `{.cast(uncheckedAssign).}` assignment to discriminator produces `[FieldDefect]`
In the manual.rst under the `cast uncheckedAssign` section there a two issues.  
1) A typo `unsafeAssign` should be `uncheckedAssign`.

2) The manual says assignment to the discriminator under `{.cast(uncheckedAssign).}` should work, but it produces: 
`Error: unhandled exception: assignment to discriminant changes object branch; compile with -d:nimOldCaseObjects for a transition period [FieldDefect]`

Here's the manual sample on nim playground: https://play.nim-lang.org/#ix=3IGq

And a smaller repro:
```nim
type
  TokenKind* = enum
    strLit, intLit
  Token = object
    case kind*: TokenKind
    of strLit:
      s*: string
    of intLit:
      i*: int64

var t = Token(kind: strLit, s: "abc")

{.cast(uncheckedAssign).}:

  # inside the 'cast' section it is allowed to assign to the 't.kind' field directly:
  t.kind = intLit
```

### Expected Output
I'd assume this should compile cleanly, or the the manual needs updating.  I don't know which would be correct.

### Additional Information
Tested on (v1.6.0) on nim playground.

### PR DESCRIPTION
**Pull request by ringabout**
fix #19266

It unblocks https://github.com/nim-lang/Nim/pull/19510

ref https://github.com/nim-lang/RFCs/issues/407

### CODE REVIEW

File: compiler/ast.nim [hunk 1]
```diff
@@ -300,6 +300,7 @@ type
     sfSingleUsedTemp  # For temporaries that we know will only be used once
     sfNoalias         # 'noalias' annotation, means C's 'restrict'
     sfEffectsDelayed  # an 'effectsDelayed' parameter
+    sfUncheckedDiscriminant
```
Comment (Araq) [inline compiler/ast.nim:303]: No...

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -1749,6 +1749,11 @@ proc semAsgn(c: PContext, n: PNode; mode=asgnNormal): PNode =
       if a.kind == nkDotCall:
         a.transitionSonsKind(nkCall)
         a = semExprWithType(c, a, {efLValue})
+    elif c.inUncheckedAssignSection > 0 and
+          a.kind in {nkDotExpr, nkCheckedFieldExpr} and
+          a[1].kind == nkSym:
+      a[1].sym.flags.incl sfUncheckedDiscriminant
```
Comment (Araq) [inline compiler/semexprs.nim:1755]: So ... when the disc field is used once in an unchecked assign section it gets tainted for good?

File: compiler/ccgstmts.nim [hunk 3]
```diff
@@ -1566,8 +1566,9 @@ proc asgnFieldDiscriminant(p: BProc, e: PNode) =
   initLocExpr(p, e[0], a)
   getTemp(p, a.t, tmp)
   expr(p, e[1], tmp)
-  if optTinyRtti notin p.config.globalOptions:
-    let field = dotExpr[1].sym
+  let field = dotExpr[1].sym
+  if optTinyRtti notin p.config.globalOptions and
+                sfUncheckedDiscriminant notin field.flags:
```
Comment (Araq) [inline compiler/ccgstmts.nim:1571]: This is not a property of the symbol! It's a property of the `.cast` construct!

File: tests/objvariant/treassign.nim [hunk 4]
```diff
@@ -34,3 +34,22 @@ proc passToVar(x: var Token) = discard
...
+        s*: string
+      of intLit:
+        i*: int64
+
+  var t = Token(kind: strLit, s: "abc")
+
+  {.cast(uncheckedAssign).}:
+
+    # inside the 'cast' section it is allowed to assign to the 't.kind' field directly:
+    t.kind = intLit
```
Comment (Araq) [inline tests/objvariant/treassign.nim:55]: And any other assignments to `t.kind` will not be checked afterwards...

**Comment by ringabout** [comment]
I agree. I will make a new PR today or tomorrow.

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/19567

**Pull request closed without merging**